src/main/java/com/google/firebase/FirebaseOptions.java

```diff
@@ -455,4 +475,27 @@ public FirebaseOptions build() {
       return new FirebaseOptions(this);
     }
   }
+
+  static class DeferredApplicationCredentials {
```

Comment:
I would expect this to directly extend the `GoogleCredentials` class. Something like:

```
// Simplified for illustration
class DeferredGoogleCredentials extends GoogleCredentials {

   private GoogleCredentials delegate;

   public AccessToken refreshAccessToken() {
       if (delegate == null) {
           delegate = GoogleCredentials.applicatioDefault();
       }
       return delegate.refreshAccessToken();
   }
}
```

Then in `FirebaseDatabase` just ignore the `GoogleCredentials` set in `FirebaseOptions` whenever connecting to the emulator.

Reply:
Alternatively, you can use that `Supplier` pattern idea proposed by @schmidt-sebastian.  `FirebaseOptions` can store credentials as a supplier:

```
private Supplier<GoogleCredentials> credentials;

// Existing public API (signature unchanged)
public Builder setGoogleCredentials(GoogleCredentials credentials) {
   this.credentials = Suppliers.ofInstance(credentials);
}

// New internal API we will use when setting up ADC
Builder setGoogleCredentials(Supplier<GoogleCredentials> supplier) {
   this.credentials = supplier;
}
```

Reply:
I like the supplier idea better. `GoogleCredentials` has way too many methods and overwriting all of them with delegate would be a lot of noise. I don't even know if it can be done cleanly since GoogleCredentials is a concrete and serializable class.

Reply:
`refreshAccessToken()` is the only method you have to override if you decide to go down that path. But, yeah, Supplier is probably a cleaner approach.

Reply:
I went with the supplier approach